overlord/ifacestate/apparmorprompting/common/common.go

```diff
@@ -0,0 +1,789 @@
+//	- /foo/bar?
+//	- /foo/bar?/
+//	# singlestars
+//	- /foo/bar/*/baz
+//	- /foo/bar/*/
+//	- /foo/bar/*/*baz
+//	- /foo/bar/*/*
+//	- /foo/bar/*
+//	- /foo/bar/*/**
+//	# glob
+//	- /foo/bar*baz
+//	- /foo/bar*/baz
+//	- /foo/bar*/baz/**
+//	- /foo/bar*/
+//	- /foo/bar*/*baz
+//	- /foo/bar*/*/baz
+//	- /foo/bar*/*/
+//	- /foo/bar*/*
+//	- /foo/bar*
+//	# doublestars
+//	- /foo/bar/**/baz
+//	- /foo/bar/**/*baz/
+//	- /foo/bar/**/*baz
+//	# terminal doublestar
+//	- /foo/bar/**/        # These are tough... usually, /foo/bar/**/ would have precedence over
+//	- /foo/bar/**/*       # precedence over /foo/bar/**/*baz, but in this case,
+//	- /foo/bar/**         # the trailing *baz adds more specificity.
+//	# glob with immediate doublestar
+//	- /foo/bar*/**/baz
+//	- /foo/bar*/**/
+//	- /foo/bar*/**
+func GetHighestPrecedencePattern(patterns []string) (string, error) {
+	if len(patterns) == 0 {
+		return "", ErrNoPatterns
+	}
+	// Map pattern to number of escaped characters which have been seen
+	remainingPatterns := make(map[string]int, len(patterns))
+	for _, pattern := range patterns {
+		remainingPatterns[pattern] = 0
+	}
+	// Loop over index into each pattern until only one pattern left
+	for i := 0; len(remainingPatterns) > 1; i++ {
+		nextPatterns := nextPatternsContainer{}
+		for pattern, e := range remainingPatterns {
+			// For each pattern, e is number of escaped chars, thus the number
+			// which should be added to the index to compare equivalent indices
+			// into all patterns
+			if i+e == len(pattern) {
+				nextPatterns.addWithPriority(priorityTerminated, pattern, e)
+				continue
+			}
+			// Check for '?' and '*' before '\\', since "\\*" is literal '*', etc.
+			if pattern[i+e] == '?' {
+				nextPatterns.addWithPriority(prioritySingleChar, pattern, e)
+				continue
+			}
+			if pattern[i+e] == '*' {
+				if i+e+3 < len(pattern) && pattern[i+e+1:i+e+4] == "/**" {
+					// Next parts of pattern are "*/**"
+					nextPatterns.addWithPriority(priorityGlobDoublestar, pattern, e)
+					continue
+				}
+				nextPatterns.addWithPriority(priorityGlob, pattern, e)
+				continue
+			}
+			if pattern[i+e] == '\\' {
+				e += 1
+				if i+e == len(pattern) {
+					return "", fmt.Errorf(`invalid path pattern: trailing '\' character: %q`, pattern)
+				}
+			}
+			// Can safely check for '/' after '\\', since it is '/' either way
+			if pattern[i+e] != '/' || i+e+1 >= len(pattern) || pattern[i+e+1] != '*' {
+				// Next parts of pattern are not "/*" or "/**"
+				nextPatterns.addWithPriority(priorityLiteral, pattern, e)
+				continue
+			}
+			// pattern[i+e:i+e+2] must be "/*"
+			if i+e+2 >= len(pattern) || pattern[i+e+2] != '*' {
+				// pattern[i+e:i+e+3] must not be "/**"
+				nextPatterns.addWithPriority(prioritySinglestar, pattern, e)
+				continue
+			}
+			// pattern[i+e:i+e+3] must be "/**"
+			if i+e+3 == len(pattern) || (pattern[i+e+3] == '/' && (i+e+4 == len(pattern) || (i+e+5 == len(pattern) && pattern[i+e+4] == '*'))) {
```

Comment:
Rework these to be careful about reading runes to look ahead. Probably read into a new reader instead, and use `ReadRune()` on that buffer to advance.